src/test/java/com/aws/greengrass/clientdevices/auth/certificate/ClientCertificateStoreTest.java

```diff
@@ -0,0 +1,115 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.io.TempDir;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+@ExtendWith({MockitoExtension.class, GGExtension.class})
+public class ClientCertificateStoreTest {
+    @TempDir
+    Path workDir;
+    private ClientCertificateStore certificateStore;
+
+    @BeforeEach
+    void beforeEach() {
+        certificateStore = new ClientCertificateStore(workDir);
+    }
+
+    @Test
+    void GIVEN_emptyCertificateStore_WHEN_createCert_THEN_certIsStoredInWorkDir() throws IOException {
+        String certPem = "pem";
+        String certId = "id";
+
+        assertThat(certificateStore.exists(certId), is(false));
+        certificateStore.storePem(certId, certPem);
+        assertThat(certificateStore.exists(certId), is(true));
+
+        Optional<String> returnedPem = certificateStore.getPem(certId);
+        assertThat(returnedPem.isPresent(), is(true));
+        assertThat(returnedPem.get(), is(certPem));
+
+        // Count files in work dir
+        try (Stream<Path> stream = Files.walk(workDir)) {
+            long count = stream.filter(Files::isRegularFile)
+                    .count();
+            assertThat(count, is(1L));
+        }
+    }
+
+    @Test
+    void GIVEN_multipleClientCertificates_WHEN_getPem_THEN_correctCertificateReturned() throws IOException {
+        Function<Integer, String> getId = (i) -> "id-" + i;
+        Function<Integer, String> getPem = (i) -> "pem-" + i;
+        long certCount = 10;
+
+        for (int i = 0; i < certCount; i++) {
+            assertThat(certificateStore.exists(getId.apply(i)), is(false));
+            certificateStore.storePem(getId.apply(i), getPem.apply(i));
+        }
+
+        for (int i = 0; i < certCount; i++) {
+            String id = getId.apply(i);
+
+            assertThat(certificateStore.exists(id), is(true));
+            Optional<String> returnedPem = certificateStore.getPem(id);
+            assertThat(returnedPem.isPresent(), is(true));
+            assertThat(returnedPem.get(), is(getPem.apply(i)));
+        }
+
+        // Count files in work dir
+        try (Stream<Path> stream = Files.walk(workDir)) {
+            long count = stream.filter(Files::isRegularFile).count();
+            assertThat(count, is(certCount));
+        }
+    }
+
+    @Test
+    void GIVEN_workDirectoryWithClientCertificates_WHEN_getPem_THEN_correctCertificateReturned() throws IOException {
+        Function<Integer, String> getId = (i) -> "id-" + i;
+        Function<Integer, String> getPem = (i) -> "pem-" + i;
+        long certCount = 10;
```

Comment:
I don't really mind having these functions but it wouldn't hurt to create real certificates with real pems and store that real data instead of fake data, given it better represents the functionality of the class.

We already have helpers for creating a root ca so you could create a method that just creates a bunch of client certificates

Reply:
I started doing that, but then talked myself out of it (time sensitivity + performance reasons - our tests are already quite slow). But I don't disagree